internal/p2p/channel.go

```diff
@@ -259,17 +264,85 @@ func (ch *libp2pChannelImpl) canonicalizedTopicName() string {
 }
 
 func (ch *libp2pChannelImpl) Receive(ctx context.Context) *ChannelIterator {
+	// TODO: consider caching an iterator in the channel, or
+	// erroring if this gets called more than once.
+	//
+	// While it's safe to register a handler more than once, we
+	// could get into a dodgy situation where if you call receive
+	// more  than once, the subsequently messages won't be routed
+	// correctly.
 	iter := &ChannelIterator{
 		pipe: make(chan Envelope),
 	}
 
+	ch.host.SetStreamHandler(protocol.ID(ch.canonicalizedTopicName()), func(stream network.Stream) {
+		buf := bufio.NewReader(stream)
+		// TODO: change the max delimited message into
+		// something that's passed in as a parmeter.
+		reader := protoio.NewDelimitedReader(buf, 1023)
+
+		remote := stream.Conn().RemotePeer()
+
+		for {
+			if ctx.Err() != nil {
+				return
+			}
+
+			payload := proto.Clone(ch.chDesc.MessageType)
+
+			// TODO: context might expire while reading
+			// the message, and so we could end up
+			// blocking here without any way to clean
+			// things up.
+			if _, err := reader.ReadMsg(payload); err != nil {
+				if errors.Is(err, io.EOF) {
+					return
+
+				}
+				// TODO: this should cause the channel to
+				// report an error, or abort the receiver.
+				_ = stream.Close()
```

Comment:
There are several paths out of this function, and the docs suggest we are obliged to call this for cleanup. Maybe defer it at the top level?

Reply:
interesting. this stream isn't one that we're creating, and instead it's passed into the handler func that we're registering here. So I'd assume that the caller of this function would be responsible for it's lifecycle. So maybe we shouldn't call close at all?

Reply:
Good question. The docs suggest we are supposed to make sure it gets called, but I didn't check if the wrapper already handles it. If so that would be even better.

Reply:
I did the thing with the context in the stream handler, and have yet to decide what the right thing to do is with the stream handler that we use from the send side.

Reply:
I did the thing that triggers the close on the context cancellation in the handler, so I think this end of things is ok. 

I think the stream we create when try and send a message will require some more finesse.